sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -87,17 +102,159 @@ private static DoFnSignature parseSignature(Class<? extends DoFn> fnClass) {
     Method setupMethod = findAnnotatedMethod(DoFn.Setup.class, fnClass, false);
     Method teardownMethod = findAnnotatedMethod(DoFn.Teardown.class, fnClass, false);
 
-    return DoFnSignature.create(
-        fnClass,
-        analyzeProcessElementMethod(fnToken, processElementMethod, inputT, outputT),
+    Method getInitialRestrictionMethod =
+        findAnnotatedMethod(DoFn.GetInitialRestriction.class, fnClass, false);
+    Method splitRestrictionMethod =
+        findAnnotatedMethod(DoFn.SplitRestriction.class, fnClass, false);
+    Method getRestrictionCoderMethod =
+        findAnnotatedMethod(DoFn.GetRestrictionCoder.class, fnClass, false);
+    Method newTrackerMethod = findAnnotatedMethod(DoFn.NewTracker.class, fnClass, false);
+
+    DoFnSignature.ProcessElementMethod processElement =
+        analyzeProcessElementMethod(fnToken, processElementMethod, inputT, outputT);
+
+    DoFnSignature.BundleMethod startBundle =
         (startBundleMethod == null)
             ? null
-            : analyzeBundleMethod(fnToken, startBundleMethod, inputT, outputT),
+            : analyzeBundleMethod(fnToken, startBundleMethod, inputT, outputT);
+    DoFnSignature.BundleMethod finishBundle =
         (finishBundleMethod == null)
             ? null
-            : analyzeBundleMethod(fnToken, finishBundleMethod, inputT, outputT),
-        (setupMethod == null) ? null : analyzeLifecycleMethod(setupMethod),
-        (teardownMethod == null) ? null : analyzeLifecycleMethod(teardownMethod));
+            : analyzeBundleMethod(fnToken, finishBundleMethod, inputT, outputT);
+    DoFnSignature.LifecycleMethod setup =
+        (setupMethod == null) ? null : analyzeLifecycleMethod(setupMethod);
+    DoFnSignature.LifecycleMethod teardown =
+        (teardownMethod == null) ? null : analyzeLifecycleMethod(teardownMethod);
+
+    DoFnSignature.GetInitialRestrictionMethod getInitialRestriction =
+        (getInitialRestrictionMethod == null)
+            ? null
+            : analyzeGetInitialRestrictionMethod(fnToken, getInitialRestrictionMethod, inputT);
+    DoFnSignature.SplitRestrictionMethod splitRestriction =
+        (splitRestrictionMethod == null)
+            ? null
+            : analyzeSplitRestrictionMethod(fnToken, splitRestrictionMethod, inputT);
+    DoFnSignature.GetRestrictionCoderMethod getRestrictionCoder =
+        (getRestrictionCoderMethod == null)
+            ? null
+            : analyzeGetRestrictionCoderMethod(fnToken, getRestrictionCoderMethod);
+    DoFnSignature.NewTrackerMethod newTracker =
+        (newTrackerMethod == null) ? null : analyzeNewTrackerMethod(fnToken, newTrackerMethod);
+
+    // Additional validation for splittable DoFn's.
+    if (processElement.isSplittable()) {
+      if (isBounded == null) {
+        isBounded =
+            processElement.hasReturnValue()
+                ? PCollection.IsBounded.UNBOUNDED
+                : PCollection.IsBounded.BOUNDED;
+      }
+      checkNotNull(
```

Comment:
It would be nice to condense these exceptions to throw once with something like

"... defines @ProcessElement method ... which is splittable, but does not define required methods %s, missingMethods"

so you get the failure once, and then can fix everything without having to run your pipeline.